libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+  {
+    if (!settings_.no_bouncer) {
+      ErrContext err{};
+      if (!query_bouncer(nettest_.name(), nettest_.test_helpers(),
+                         nettest_.version(), &ctx.collectors,
+                         &ctx.test_helpers, &info, &err)) {
+        LIBNETTEST2_EMIT_WARNING("run: query_bouncer() failed");
+        // TODO(bassosimone): shouldn't we introduce failure.query_bouncer?
+        // FALLTHROUGH
+      }
+    }
+  }
+  emit_ev("status.progress", {{"percentage", 0.1},
+                              {"message", "contact bouncer"}});
+  {
+    if (settings_.probe_ip == "") {
+      if (!settings_.no_ip_lookup) {
+        ErrContext err{};
+        if (!lookup_ip(&ctx.probe_ip, &info, &err)) {
+          LIBNETTEST2_EMIT_WARNING("run: lookup_ip() failed");
+          emit_ev("failure.ip_lookup", {
+              {"failure", "library_error"},
+              {"library_error_context", err},
+          });
+        }
+      }
+    } else {
+      ctx.probe_ip = settings_.probe_ip;
+    }
+    LIBNETTEST2_EMIT_DEBUG("probe_ip: " << ctx.probe_ip);
+  }
+  {
+    // Implementation detail: if probe_asn is empty then we will also overwrite
+    // the value inside of probe_network_name even if it's non-empty.
+    if (settings_.probe_asn == "") {
+      if (!settings_.no_asn_lookup) {
+        ErrContext err{};
+        if (!lookup_asn(settings_.geoip_asn_path, ctx.probe_ip, &ctx.probe_asn,
+                        &ctx.probe_network_name, &err)) {
+          LIBNETTEST2_EMIT_WARNING("run: lookup_asn() failed");
+          emit_ev("failure.asn_lookup", {
+              {"failure", "library_error"},
+              {"library_error_context", err},
+          });
+        }
+      }
+    } else {
+      ctx.probe_network_name = settings_.probe_network_name;
+      ctx.probe_asn = settings_.probe_asn;
+    }
+    LIBNETTEST2_EMIT_DEBUG("probe_asn: " << ctx.probe_asn);
+    LIBNETTEST2_EMIT_DEBUG("probe_network_name: " << ctx.probe_network_name);
+  }
+  {
+    if (settings_.probe_cc == "") {
+      if (!settings_.no_cc_lookup) {
+        ErrContext err{};
+        if (!lookup_cc(settings_.geoip_country_path, ctx.probe_ip,
+                       &ctx.probe_cc, &err)) {
+          LIBNETTEST2_EMIT_WARNING("run: lookup_cc() failed");
+          emit_ev("failure.cc_lookup", {
+              {"failure", "library_error"},
+              {"library_error_context", err},
+          });
+        }
+      }
+    } else {
+      ctx.probe_cc = settings_.probe_cc;
+    }
+    LIBNETTEST2_EMIT_DEBUG("probe_cc: " << ctx.probe_cc);
+  }
+  emit_ev("status.progress", {{"percentage", 0.2},
+                              {"message", "geoip lookup"}});
+  // TODO(bassosimone): make sure that passing empty strings here is
+  // preferrable than passing conventional values like 127.0.0.1. This
+  // has been prompted by a discussion with @lorenzoPrimi and I guess
+  // it needs to be also discussed with @hellais.
```

Comment:
This seems like a reasonable suggestion.

Reply:
Great, then I will make sure the spec is also consistent.

Reply:
See ae9cb4b2ebcc7f6774f691d8429a4148c6fac8b9